Error with pytest-playwright 0.6

Minimal reproducer:

```
python -m venv clean
source clean/bin/activate
pip install solara[pytest]
```

then create ``test.py`` with:

```python
def test_basic(page_session):
    pass
```

And run the tests with:

```
pytest test.py
```

this fails with:

```
    @pytest.fixture(scope="session")
    def context_session(
        browser: "playwright.sync_api.Browser",
        browser_context_args: Dict,
        pytestconfig: Any,
        request: pytest.FixtureRequest,
    ) -> Generator["playwright.sync_api.BrowserContext", None, None]:
        from playwright.sync_api import Error, Page
>       from pytest_playwright.pytest_playwright import _build_artifact_test_folder
E       ImportError: cannot import name '_build_artifact_test_folder' from 'pytest_playwright.pytest_playwright' (/home/tom/tmp/debug/clean/lib/python3.11/site-packages/pytest_playwright/pytest_playwright.py)

clean/lib/python3.11/site-packages/solara/test/pytest_plugin.py:66: ImportError
```

Downgrading to pytest-playwright 0.5 fixes this. It looks like some private API was removed?

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -284,7 +284,7 @@ jobs:
           pip install `echo packages/solara-server/dist/*.whl`[all]
           pip install `echo packages/solara-meta/dist/*.whl`[dev,documentation]
           # pefile >= 2023.2.7 make pyinstaller incredibly slow https://github.com/erocarrera/pefile/issues/420
-          pip install "jupyterlab<4"  "pydantic<2" "playwright==1.41.2" pyinstaller pefile==2023.2.7
+          pip install "jupyterlab<4"  "pydantic<2" "playwright==1.50.0" pyinstaller pefile==2023.2.7
           pip freeze --exclude solara --exclude solara-ui --exclude solara-server > ${{ env.LOCK_FILE_LOCATION }}
           git diff --quiet || echo "HAS_DIFF=true" >> "$GITHUB_OUTPUT"
           git diff | tee ${{ env.DIFF_FILE_LOCATION }}
@@ -403,7 +403,7 @@ jobs:
           pip install `echo packages/pytest-ipywidgets/dist/*.whl`[all,ipywidgets${{ matrix.ipywidgets_major }}]
           pip install `echo packages/solara-meta/dist/*.whl`[dev,documentation]
           pip install `echo packages/solara-enterprise/dist/*.whl`[ssg,auth]
-          pip install "jupyterlab<4"  "pydantic<2" "playwright==1.41.2" "ipywidgets~=${{ matrix.ipywidgets }}"
+          pip install "jupyterlab<4"  "pydantic<2" "playwright==1.50.0" "ipywidgets~=${{ matrix.ipywidgets }}"
           pip freeze --exclude solara --exclude solara-ui --exclude solara-server --exclude pytest-ipywidgets --exclude solara-enterprise > ${{ env.LOCK_FILE_LOCATION }}
           git diff | tee ${{ env.DIFF_FILE_LOCATION }}
           [ -s ${{ env.DIFF_FILE_LOCATION }} ] && echo "HAS_DIFF=true" >> "$GITHUB_OUTPUT" || echo "No dependencies changed"
@@ -510,7 +510,7 @@ jobs:
           pip install `echo packages/pytest-ipywidgets/dist/*.whl`[all,ipywidgets${{ matrix.ipywidgets_major }}]
           pip install `echo packages/solara-meta/dist/*.whl`[dev,documentation]
           pip install `echo packages/solara-enterprise/dist/*.whl`[ssg,auth]
-          pip install "jupyterlab<4" "playwright==1.41.2" "pydantic<2" "ipywidgets~=${{ matrix.ipywidgets }}"
+          pip install "jupyterlab<4" "playwright==1.50.0" "pydantic<2" "ipywidgets~=${{ matrix.ipywidgets }}"
           pip install jupyter_core jupyter-packaging
           pip install --pre "ipyvue>=3" "ipyvuetify>=3"
           pip freeze --exclude solara --exclude solara-ui --exclude solara-server --exclude pytest-ipywidgets --exclude solara-enterprise > ${{ env.LOCK_FILE_LOCATION }}

--- a/solara/test/pytest_plugin.py
+++ b/solara/test/pytest_plugin.py
@@ -9,11 +9,12 @@ import sys
 import textwrap
 import threading
 import typing
+import tempfile
 import urllib.parse
 import uuid
 from io import BytesIO
 from pathlib import Path
-from typing import Any, Callable, Dict, Generator, List, Union
+from typing import Any, Callable, Dict, Generator, List, Union, cast
 
 import ipywidgets as widgets
 import pytest
@@ -55,74 +56,143 @@ def solara_server(pytestconfig: Any, request):
         webserver.stop_serving()
 
 
-@pytest.fixture(scope="session")
-def context_session(
-    browser: "playwright.sync_api.Browser",
-    browser_context_args: Dict,
-    pytestconfig: Any,
-    request: pytest.FixtureRequest,
-) -> Generator["playwright.sync_api.BrowserContext", None, None]:
-    from playwright.sync_api import Error, Page
-    from pytest_playwright.pytest_playwright import _build_artifact_test_folder
-    from slugify import slugify  # type: ignore
+try:
+    from pytest_playwright.pytest_playwright import connect_options  # noqa
 
-    pages: List[Page] = []
-    context = browser.new_context(**browser_context_args)
-    context.on("page", lambda page: pages.append(page))
+    has_pytest_070 = True
+except ImportError:
+    has_pytest_070 = False
 
-    tracing_option = pytestconfig.getoption("--tracing")
-    capture_trace = tracing_option in ["on", "retain-on-failure"]
-    if capture_trace:
-        context.tracing.start(
-            title=slugify(request.node.nodeid),
-            screenshots=True,
-            snapshots=True,
-            sources=True,
-        )
 
-    # decrease timeout from 30s to 10s
-    context.set_default_timeout(TIMEOUT * 1000)
-    yield context
+if has_pytest_070:
+    # pytest-playwright >= 0.7.0
+    from pytest_playwright.pytest_playwright import Browser, CreateContextCallback, BrowserContext, Playwright, ArtifactsRecorder, _truncate_file_name
+    from slugify import slugify
 
-    # If request.node is missing rep_call, then some error happened during execution
-    # that prevented teardown, but should still be counted as a failure
-    failed = request.node.rep_call.failed if hasattr(request.node, "rep_call") else True
+    @pytest.fixture(scope="session")
+    def output_path_session(pytestconfig: Any, request: pytest.FixtureRequest) -> str:
+        output_dir = Path(pytestconfig.getoption("--output")).absolute()
+        return os.path.join(output_dir, _truncate_file_name(slugify(request.node.nodeid)))
 
-    if capture_trace:
-        retain_trace = tracing_option == "on" or (failed and tracing_option == "retain-on-failure")
-        if retain_trace:
-            trace_path = _build_artifact_test_folder(pytestconfig, request, "trace.zip")
-            context.tracing.stop(path=trace_path)
-        else:
-            context.tracing.stop()
+    @pytest.fixture(scope="session")
+    def _artifacts_recorder_session(
+        request: pytest.FixtureRequest,
+        output_path_session: str,
+        playwright: Playwright,
+        pytestconfig: Any,
+        _pw_artifacts_folder: tempfile.TemporaryDirectory,
+    ) -> Generator["ArtifactsRecorder", None, None]:
+        artifacts_recorder = ArtifactsRecorder(pytestconfig, request, output_path_session, playwright, _pw_artifacts_folder)
+        yield artifacts_recorder
+        # If request.node is missing rep_call, then some error happened during execution
+        # that prevented teardown, but should still be counted as a failure
+        failed = request.node.rep_call.failed if hasattr(request.node, "rep_call") else True
+        artifacts_recorder.did_finish_test(failed)
 
-    screenshot_option = pytestconfig.getoption("--screenshot")
-    capture_screenshot = screenshot_option == "on" or (failed and screenshot_option == "only-on-failure")
-    if capture_screenshot:
-        for index, page in enumerate(pages):
-            human_readable_status = "failed" if failed else "finished"
-            screenshot_path = _build_artifact_test_folder(pytestconfig, request, f"test-{human_readable_status}-{index + 1}.png")
-            try:
-                page.screenshot(timeout=5000, path=screenshot_path)
-            except Error:
-                pass
+    @pytest.fixture(scope="session")
+    def new_context_session(
+        browser: Browser,
+        browser_context_args: Dict,
+        _artifacts_recorder_session: "ArtifactsRecorder",
+        request: pytest.FixtureRequest,
+    ) -> Generator[CreateContextCallback, None, None]:
+        browser_context_args = browser_context_args.copy()
+        context_args_marker = next(request.node.iter_markers("browser_context_args"), None)
+        additional_context_args = context_args_marker.kwargs if context_args_marker else {}
+        browser_context_args.update(additional_context_args)
+        contexts: List[BrowserContext] = []
 
-    context.close()
+        def _new_context(**kwargs: Any) -> BrowserContext:
+            context = browser.new_context(**browser_context_args, **kwargs)
+            original_close = context.close
 
-    video_option = pytestconfig.getoption("--video")
-    preserve_video = video_option == "on" or (failed and video_option == "retain-on-failure")
-    if preserve_video:
-        for page in pages:
-            video = page.video
-            if not video:
-                continue
-            try:
-                video_path = video.path()
-                file_name = os.path.basename(video_path)
-                video.save_as(path=_build_artifact_test_folder(pytestconfig, request, file_name))
-            except Error:
-                # Silent catch empty videos.
-                pass
+            def _close_wrapper(*args: Any, **kwargs: Any) -> None:
+                contexts.remove(context)
+                _artifacts_recorder_session.on_will_close_browser_context(context)
+                original_close(*args, **kwargs)
+
+            context.close = _close_wrapper
+            contexts.append(context)
+            _artifacts_recorder_session.on_did_create_browser_context(context)
+            return context
+
+        yield cast(CreateContextCallback, _new_context)
+        for context in contexts.copy():
+            context.close()
+
+    @pytest.fixture(scope="session")
+    def context_session(new_context_session):
+        return new_context_session()
+else:
+
+    @pytest.fixture(scope="session")
+    def context_session(
+        browser: "playwright.sync_api.Browser",
+        browser_context_args: Dict,
+        pytestconfig: Any,
+        request: pytest.FixtureRequest,
+    ) -> Generator["playwright.sync_api.BrowserContext", None, None]:
+        from playwright.sync_api import Error, Page
+        from pytest_playwright.pytest_playwright import _build_artifact_test_folder
+        from slugify import slugify  # type: ignore
+
+        pages: List[Page] = []
+        context = browser.new_context(**browser_context_args)
+        context.on("page", lambda page: pages.append(page))
+
+        tracing_option = pytestconfig.getoption("--tracing")
+        capture_trace = tracing_option in ["on", "retain-on-failure"]
+        if capture_trace:
+            context.tracing.start(
+                title=slugify(request.node.nodeid),
+                screenshots=True,
+                snapshots=True,
+                sources=True,
+            )
+
+        # decrease timeout from 30s to 10s
+        context.set_default_timeout(TIMEOUT * 1000)
+        yield context
+
+        # If request.node is missing rep_call, then some error happened during execution
+        # that prevented teardown, but should still be counted as a failure
+        failed = request.node.rep_call.failed if hasattr(request.node, "rep_call") else True
+
+        if capture_trace:
+            retain_trace = tracing_option == "on" or (failed and tracing_option == "retain-on-failure")
+            if retain_trace:
+                trace_path = _build_artifact_test_folder(pytestconfig, request, "trace.zip")
+                context.tracing.stop(path=trace_path)
+            else:
+                context.tracing.stop()
+
+        screenshot_option = pytestconfig.getoption("--screenshot")
+        capture_screenshot = screenshot_option == "on" or (failed and screenshot_option == "only-on-failure")
+        if capture_screenshot:
+            for index, page in enumerate(pages):
+                human_readable_status = "failed" if failed else "finished"
+                screenshot_path = _build_artifact_test_folder(pytestconfig, request, f"test-{human_readable_status}-{index + 1}.png")
+                try:
+                    page.screenshot(timeout=5000, path=screenshot_path)
+                except Error:
+                    pass
+
+        context.close()
+
+        video_option = pytestconfig.getoption("--video")
+        preserve_video = video_option == "on" or (failed and video_option == "retain-on-failure")
+        if preserve_video:
+            for page in pages:
+                video = page.video
+                if not video:
+                    continue
+                try:
+                    video_path = video.path()
+                    file_name = os.path.basename(video_path)
+                    video.save_as(path=_build_artifact_test_folder(pytestconfig, request, file_name))
+                except Error:
+                    # Silent catch empty videos.
+                    pass
 
 
 # page fixture that keeps open all the time, is faster

